src/target/riscv/riscv-013.c

```diff
@@ -2001,6 +2001,226 @@ static int riscv013_set_register_buf(struct target *target,
+		return ERROR_NOT_IMPLEMENTED;
+	}
+
+	if (get_field(info->sbcs, DM_SBCS_SBVERSION) != 1) {
+		LOG_ERROR("Memory sampling is only implemented for SBA version 1.");
+		return ERROR_NOT_IMPLEMENTED;
+	}
+
+	uint32_t sbcs = 0;
+	uint32_t sbcs_valid = false;
+
+	uint32_t sbaddress0 = 0;
+	bool sbaddress0_valid = false;
+	uint32_t sbaddress1 = 0;
+	bool sbaddress1_valid = false;
+
+	unsigned enabled_count = 0;
+	for (unsigned i = 0; i < DIM(config->bucket); i++) {
+		if (config->bucket[i].enabled)
+			enabled_count++;
+	}
+
+	while (timeval_ms() < until_ms) {
+		/* 
+		 * batch_run() adds to the batch, so we can't simply reuse the same
+		 * batch over and over. So we create a new one every time through the
+		 * loop. *
+		 */
+		struct riscv_batch *batch = riscv_batch_alloc(
+			target, DIM(config->bucket) * 6,
+			info->dmi_busy_delay + info->bus_master_read_delay);
+
+		unsigned result_bytes = 0;
+		for (unsigned i = 0; i < DIM(config->bucket); i++) {
+			if (config->bucket[i].enabled) {
+				uint32_t sbaccess = 0;
+				switch (config->bucket[i].size_bytes) {
+					case 1:
+						sbaccess = DM_SBCS_SBACCESS8;
+						break;
+					case 2:
+						sbaccess = DM_SBCS_SBACCESS16;
+						break;
+					case 4:
+						sbaccess = DM_SBCS_SBACCESS32;
+						break;
+					case 8:
+						sbaccess = DM_SBCS_SBACCESS64;
+						break;
+					default:
+						LOG_ERROR("Unsupported SBA access size: %d",
+								  config->bucket[i].size_bytes);
+						return ERROR_NOT_IMPLEMENTED;
+				}
+				if (!(info->sbcs & sbaccess)) {
+					LOG_ERROR("Hardware does not support SBA access for %d-byte memory sampling.",
+							  config->bucket[i].size_bytes);
+					return ERROR_NOT_IMPLEMENTED;
+				}
+
+				uint32_t sbcs_write = DM_SBCS_SBREADONADDR;
+				if (enabled_count == 1)
+					sbcs_write |= DM_SBCS_SBREADONDATA;
+				sbcs_write |= sb_sbaccess(config->bucket[i].size_bytes);
+				if (!sbcs_valid || sbcs_write != sbcs) {
+					riscv_batch_add_dmi_write(batch, DM_SBCS, sbcs_write);
+					sbcs = sbcs_write;
+					sbcs_valid = true;
+				}
+
+				if (sbasize > 32 &&
+						(!sbaddress1_valid ||
+						 sbaddress1 != config->bucket[i].address >> 32)) {
+					sbaddress1 = config->bucket[i].address >> 32;
+					riscv_batch_add_dmi_write(batch, DM_SBADDRESS1, sbaddress1);
+					sbaddress1_valid = true;
+				}
+				if (!sbaddress0_valid ||
+						sbaddress0 != (config->bucket[i].address & 0xffffffff)) {
+					sbaddress0 = config->bucket[i].address;
+					riscv_batch_add_dmi_write(batch, DM_SBADDRESS0, sbaddress0);
+					sbaddress0_valid = true;
+				}
+				if (config->bucket[i].size_bytes > 4)
+					riscv_batch_add_dmi_read(batch, DM_SBDATA1);
+				riscv_batch_add_dmi_read(batch, DM_SBDATA0);
+				result_bytes += 1 + config->bucket[i].size_bytes;
+			}
+		}
+
+		if (buf->used + result_bytes >= buf->size) {
+			riscv_batch_free(batch);
+			break;
+		}
+
+		/*int result =*/ batch_run(target, batch);
+		unsigned read = 0;
```

Comment:
After the batch finishes, please consider checking for the possible error states: `sbcs.sberror`, `sbcs.sbbusyerror`. If they occur, the user should be notified, and the error state cleared.

This is important because subsequent system bus accesses (e.g. `read_memory_bus_v1()`) assume that any previous sticky errors had been cleared.

Reply:
I knew there was something I forgot! Done.